File: python-solid-principle-sample-code/com/principle/solid/openclosed/AreaCalculator.py
from math import pi
from typing import List

from Shape import Shape
from Circle import Circle
from Square import Square
from Cube import Cube
from Rectangle import Rectangle


class AreaCalculator:

    def sum(self, shapes: List[Shape]):
        total_area = 0
        for shape in shapes:
            # if isinstance(shape, Square):
            #     total_area += shape.get_length() ** 2
            # if isinstance(shape, Circle):
            #     total_area += pi * shape.get_radius() ** 2

            """
                Open Closed Principle violation - It is Open for Modification but it should be Open for Extension
            """
            # if isinstance(shape, Cube):
            #     total_area += 6 * shape.get_length() ** 2

            """
                Open Closed Principle Achieved - Using Interface
            """
            total_area += shape.area()

        return total_area

    """
        single responsibility principle violation
    """
    # def json(self, shapes):
    #     return "{{sum: {}}}".format(int(self.sum(shapes)))

File: java-solid-principle-sample-code/src/com/principle/solid/liskov_substitution/AreaCalculator.java
package com.principle.solid.liskov_substitution;

import java.util.List;

public class AreaCalculator {

	public int sum(List<Shape> shapes) {

		int sum = 0;
		for(int i = 0; i < shapes.size(); i++) {
			Shape shape = shapes.get(i);
//			if(shape instanceof Square) {
//				sum += Math.pow(((Square)shape).getLength(), 2);
//			}
//			if(shape instanceof Circle) {
//				sum += Math.PI * Math.pow(((Circle)shape).getRadius(), 2);
//			}
			/** Open Closed Principle violation - It is Open for Modification but it should be Open for Extension */
//			if(shape instanceof Cube) {
//				sum += Math.PI * Math.pow(((Circle)shape).getRadius(), 2);
//			}
			
			/** Open Closed Principle Achieved - Using Interface */
			sum += shape.area();
		}
		return sum;
	}
	
	/** single responsibility principle violation */
//	public String json(List<Shape> shapes) {
//		return String.format("{sum: %s}", sum(shapes));
//	}
}
